src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -472,136 +518,192 @@ private void doSignal(
+            doData(ping.buffer(), ping.offset(), ping.sizeof());
         }
 
-        private void doMqttSuback()
+        private void doMqttSuback(
+            int subscriptions)
         {
+            OctetsFW reasonCodes = new OctetsFW.Builder()
+                .wrap(writeBuffer, 0, writeBuffer.capacity())
+                .put(new byte[] {0x00})
+                .build();
+
+            final MqttSubackFW suback = mqttSubackRW.wrap(writeBuffer, DataFW.FIELD_OFFSET_PAYLOAD, writeBuffer.capacity())
+                .remainingLength(reasonCodes.sizeof() + 1)
+                .propertiesLength(0x00)
+                .reasonCodes(reasonCodes)
+                .build();
+
+            doData(suback.buffer(), suback.offset(), suback.sizeof());
         }
 
-        private void doMqttUnsuback()
+        private void doMqttUnsuback(
+            int subsriptions)
         {
+            OctetsFW reasonCodes = new OctetsFW.Builder()
+                .wrap(writeBuffer, 0, writeBuffer.capacity())
+                .put(new byte[] {0x00})
+                .build();
+
+            final MqttUnsubackFW unsuback = mqttUnsubackRW.wrap(writeBuffer, DataFW.FIELD_OFFSET_PAYLOAD, writeBuffer.capacity())
+                .remainingLength(reasonCodes.sizeof() + 1)
+                .propertiesLength(0x00)
+                .reasonCodes(reasonCodes)
+                .build();
+
+            doData(unsuback.buffer(), unsuback.offset(), unsuback.sizeof());
         }
 
-        private void doMqttDisconnect()
+        private void doMqttDisconnect(
+            int reasonCode)
         {
             OctetsFW properties = new OctetsFW.Builder()
                 .wrap(writeBuffer, 0, writeBuffer.capacity())
                 .build();
 
             final MqttDisconnectFW disconnect = mqttDisconnectRW
                 .wrap(writeBuffer,  DataFW.FIELD_OFFSET_PAYLOAD, writeBuffer.capacity())
-                .remainingLength(0x00)
-                .reasonCode(0x00)
+                .remainingLength((byte) properties.sizeof() + 1)
+                .reasonCode(reasonCode)
                 .properties(properties)
                 .build();
 
-            final DataFW data = dataRW.wrap(writeBuffer, 0, writeBuffer.capacity())
-                .routeId(routeId)
-                .streamId(replyId)
-                .trace(supplyTraceId.getAsLong())
-                .groupId(0)
-                .padding(replyPadding)
-                .payload(disconnect.buffer(), disconnect.offset(), disconnect.sizeof())
-                .build();
+            doData(disconnect.buffer(), disconnect.offset(), disconnect.sizeof());
+        }
 
-            receiver.accept(data.typeId(), data.buffer(), data.offset(), data.sizeof());
+        private boolean setReservedSubscriptionOptionsBits(
+            int options)
+        {
+            return ((options >> 8) & 1) != 0
+                && ((options >> 7) & 1) != 0;
         }
```

Comment:
Recommend you change this logic to verify correctness, rather than checking for failure.
Also, if this will only be validated in `onMqttSubscribe` then no need to separate it out into another method.